hal/api/critical.h

```diff
@@ -222,52 +224,121 @@ bool core_util_atomic_cas_u16(uint16_t *ptr, uint16_t *expectedCurrentValue, uin
+ *                              failure case of atomic_cas (where the
+ *                              destination isn't set), the pointee of expectedCurrentValue is
+ *                              updated with the current value.
+ * @param[in] desiredValue      The new value computed based on '*expectedCurrentValue'.
+ *
+ * @return                      true if the memory location was atomically
+ *                              updated with the desired value (after verifying
+ *                              that it contained the expectedCurrentValue),
+ *                              false otherwise. In the failure case,
+ *                              exepctedCurrentValue is updated with the new
+ *                              value of the target memory location.
+ *
+ * pseudocode:
+ * function cas(p : pointer to int, old : pointer to int, new : int) returns bool {
+ *     if *p != *old {
+ *         *old = *p
+ *         return false
+ *     }
+ *     *p = new
+ *     return true
+ * }
+ *
+ * @Note: In the failure case (where the destination isn't set), the value
+ * pointed to by expectedCurrentValue is still updated with the current value.
+ * This property helps writing concise code for the following incr:
+ *
+ * function incr(p : pointer to int, a : int) returns int {
+ *     done = false
+ *     value = *p // This fetch operation need not be atomic.
+ *     while not done {
+ *         done = atomic_cas(p, &value, value + a) // *value gets updated automatically until success
+ *     }
+ *     return value + a
+ * }
+ */
+bool core_util_atomic_cas(void **ptr, void **expectedCurrentValue, void *desiredValue);
+
+/**
+ * Atomic increment.
+ * @param  valuePtr Target memory location being incremented.
+ * @param  delta    The amount being incremented.
+ * @return          The new incremented value.
+ */
+uint8_t core_util_atomic_incr_u8(uint8_t *valuePtr, uint8_t delta);
+
+/**
  * Atomic increment.
  * @param  valuePtr Target memory location being incremented.
  * @param  delta    The amount being incremented.
  * @return          The new incremented value.
  */
-uint8_t core_util_atomic_incr_u8(uint8_t * valuePtr, uint8_t delta);
+uint16_t core_util_atomic_incr_u16(uint16_t *valuePtr, uint16_t delta);
 
 /**
  * Atomic increment.
  * @param  valuePtr Target memory location being incremented.
  * @param  delta    The amount being incremented.
  * @return          The new incremented value.
  */
-uint16_t core_util_atomic_incr_u16(uint16_t * valuePtr, uint16_t delta);
+uint32_t core_util_atomic_incr_u32(uint32_t *valuePtr, uint32_t delta);
 
 /**
  * Atomic increment.
  * @param  valuePtr Target memory location being incremented.
  * @param  delta    The amount being incremented.
  * @return          The new incremented value.
  */
-uint32_t core_util_atomic_incr_u32(uint32_t * valuePtr, uint32_t delta);
+void* core_util_atomic_incr(void **valuePtr, ptrdiff_t delta);
```

Comment:
shouldn't this function's name be suffixed with something to indicate that it works with void pointers?

Quite likely, users of this API will want to pass in pointers of some particular type, and would perhaps expect pointer arithmetic. The contract of this API should state explicitly what the incr would do to the pointer.